servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SetDynamicCompositeCancellable.java

```diff
@@ -17,75 +17,64 @@
-import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 
-import static io.servicetalk.concurrent.internal.ConcurrentUtils.CONCURRENT_EMITTING;
-import static io.servicetalk.concurrent.internal.ConcurrentUtils.CONCURRENT_IDLE;
-import static io.servicetalk.concurrent.internal.ConcurrentUtils.drainSingleConsumerCollectionDelayThrow;
+import static io.servicetalk.utils.internal.PlatformDependent.throwException;
+import static java.util.Collections.newSetFromMap;
 
 /**
  * A {@link Cancellable} that contains other {@link Cancellable}s.
  * <p>
  * {@link Cancellable#cancel()} is propagated to all active {@link Cancellable}s. Any new {@link Cancellable} added
  * after that will be immediately cancelled.
  */
-final class MapDynamicCompositeCancellable implements DynamicCompositeCancellable {
-
-    private static final AtomicIntegerFieldUpdater<MapDynamicCompositeCancellable> cancelledUpdater =
-            AtomicIntegerFieldUpdater.newUpdater(MapDynamicCompositeCancellable.class, "cancelled");
+final class SetDynamicCompositeCancellable implements DynamicCompositeCancellable {
+    private static final AtomicIntegerFieldUpdater<SetDynamicCompositeCancellable> cancelledUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(SetDynamicCompositeCancellable.class, "cancelled");
     @SuppressWarnings("unused")
     private volatile int cancelled;
 
-    private static final AtomicIntegerFieldUpdater<MapDynamicCompositeCancellable> drainingUpdater =
-            AtomicIntegerFieldUpdater.newUpdater(MapDynamicCompositeCancellable.class, "draining");
-    @SuppressWarnings("unused")
-    private volatile int draining;
-
-    private final ConcurrentMap<Cancellable, Cancellable> cancellables;
+    private final Set<Cancellable> cancellables;
 
     /**
      * Create a new instance.
      */
-    MapDynamicCompositeCancellable() {
+    SetDynamicCompositeCancellable() {
         this(8);
     }
 
     /**
      * Create a new instance.
-     * @param initialSize The initial size of the internal {@link Map}.
+     * @param initialSize The initial size of the internal {@link Set}.
      */
-    MapDynamicCompositeCancellable(int initialSize) {
-        cancellables = new ConcurrentHashMap<>(initialSize);
+    SetDynamicCompositeCancellable(int initialSize) {
+        cancellables = newSetFromMap(new ConcurrentHashMap<>(initialSize));
     }
 
     @Override
     public void cancel() {
-        if (cancelledUpdater.compareAndSet(this, CONCURRENT_IDLE, CONCURRENT_EMITTING)) {
+        if (cancelledUpdater.compareAndSet(this, 0, 1)) {
             cancelAll();
         }
     }
 
     @Override
     public boolean add(Cancellable toAdd) {
-        if (isCancelled()) {
-            toAdd.cancel();
+        if (!cancellables.add(toAdd)) {
+            toAdd.cancel(); // out of memory, or user has implemented equals/hashCode so there is overlap.
+        } else if (isCancelled()) {
+            cancelAll();
```

Comment:
Can we avoid calling `cancelAll()` here as I suggested above for the queue variant? 

Looks like we can concurrently invoke a `Cancellable` if `cancelAll()` is invoked concurrently.

Reply:
good call ... because of the `Set` interface we can't safely drain from multiple threads and process and item exactly once. This needs to use the `tryAcquireLock` method and I will fix.